### PR TITLE
[SPARKR-225] Merge master into sparkr-sql branch

### PR DESCRIPTION
**Pull request by davies**
There is lots of conflicts, fixed manually line by line, hopefully they are merged correctly.

### CODE REVIEW

**Comment by shivaram** [comment]
@davies  Thanks. I'll try to review this in detail to make sure the merge is fine.

@cafreeman  @sun-rui  -- This could probably use a couple of more people looking at it too.

**Comment by shivaram** [comment]
Actually one thing before I get into the review. It looks like this change also include the toDF changes from #212 -- Can you exclude them ? One simple way to sanity check is to make sure `DataFrame.R` is not a part of this PR

**Comment by davies** [comment]
@shivaram Once you merge #212 , I will update this PR (hopefully it will gone automatically), you can ignore it now.

File: pkg/R/RDD.R [hunk 1]
```diff
@@ -868,20 +756,15 @@ setMethod("first",
...
 #'}
-setGeneric("distinct",
-           function(x, numPartitions) { standardGeneric("distinct") })
-
 setClassUnion("missingOrInteger", c("missing", "integer"))
 #' @rdname distinct
-#' @aliases distinct,RDD,missingOrInteger-method
+#' @aliases distinct,RDD-method
 setMethod("distinct",
           signature(x = "RDD", numPartitions = "missingOrInteger"),
```
Comment (shivaram) [inline pkg/R/RDD.R]: In the master branch the default value here is set to `numPartitions = SparkR::numPartitions(x)`

File: pkg/R/RDD.R [hunk 2]
```diff
@@ -1377,8 +1282,243 @@ setMethod("setName",
             x
           })
 
+#' @examples
```
Comment (shivaram) [inline pkg/R/RDD.R]: top, fold, aggregate, pipe, name, setName are already present in RDD.R in the sparkr-sql branch. The only things to preserve are `zipWithUniqueId` and `zipWithIndex`

File: pkg/R/pairRDD.R [hunk 3]
```diff
@@ -642,12 +560,10 @@ setMethod("join",
 #' leftOuterJoin(rdd1, rdd2, 2L)
 #' # list(list(1, list(1, 2)), list(1, list(1, 3)), list(2, list(4, NULL)))
 #'}
-setGeneric("leftOuterJoin", function(x, y, numPartitions) { standardGeneric("leftOuterJoin") })
-
 #' @rdname join-methods
 #' @aliases leftOuterJoin,RDD,RDD-method
 setMethod("leftOuterJoin",
-          signature(x = "RDD", y = "RDD", numPartitions = "integer"),
```
Comment (shivaram) [inline pkg/R/pairRDD.R]: I think `numPartitions` should be preserved here

File: pkg/R/pairRDD.R [hunk 4]
```diff
@@ -683,12 +598,10 @@ setMethod("leftOuterJoin",
 #' rightOuterJoin(rdd1, rdd2, 2L)
 #' # list(list(1, list(2, 1)), list(1, list(3, 1)), list(2, list(NULL, 4)))
 #'}
-setGeneric("rightOuterJoin", function(x, y, numPartitions) { standardGeneric("rightOuterJoin") })
-
 #' @rdname join-methods
 #' @aliases rightOuterJoin,RDD,RDD-method
 setMethod("rightOuterJoin",
-          signature(x = "RDD", y = "RDD", numPartitions = "integer"),
+          signature(x = "RDD", y = "RDD"),
```
Comment (shivaram) [inline pkg/R/pairRDD.R]: same as before. we should keep `numPartitions`

File: pkg/R/pairRDD.R [hunk 5]
```diff
@@ -727,13 +639,10 @@ setMethod("rightOuterJoin",
...
 #'                               #      list(3, list(3, NULL)),
 #'}
-setGeneric("fullOuterJoin", function(x, y, numPartitions) { standardGeneric("fullOuterJoin") })
-
 #' @rdname join-methods
 #' @aliases fullOuterJoin,RDD,RDD-method
-
 setMethod("fullOuterJoin",
-          signature(x = "RDD", y = "RDD", numPartitions = "integer"),
+          signature(x = "RDD", y = "RDD"),
```
Comment (shivaram) [inline pkg/R/pairRDD.R]: here as well

File: pkg/R/pairRDD.R [hunk 6]
```diff
@@ -742,7 +651,8 @@ setMethod("fullOuterJoin",
               joinTaggedList(v, list(TRUE, TRUE))
             }
 
-            joined <- flatMapValues(groupByKey(unionRDD(xTagged, yTagged), numPartitions), doJoin)
+            joined <- flatMapValues(groupByKey(unionRDD(xTagged, yTagged), numToInt(numPartitions)),
```
Comment (shivaram) [inline pkg/R/pairRDD.R]: while this change to using `numToInt` is a good one, lets not do it in the merge PR. we can do it in a follow up PR

**Comment by shivaram** [comment]
@davies I did one pass and left some comments. I'll do another pass after merging #212 as that'll give a better idea of the diff

**Comment by shivaram** [comment]
Thanks for the update. I took another detailed look and the change looks pretty good to me. I'm going to leave this open till tomorrow to give @sun-rui or @hlin09 a chance to see if they have any comments -- but don't worry I won't merge anything into sparkr-sql before this !

**Comment by sun-rui** [comment]
LGTM

**Comment by shivaram** [comment]
Merging this.

**Pull request merged**